Desktop / MNTM Settings: Directories and generic files support for Keybinds / Main Menu

956MB opened the pull request:
# What's new

Adds *RIGHT* button select in the file browser dialogs and changes the `Open File` action to `Open File/Directory` in `Settings > Desktop > Keybinds Setup`. This adds the ability to open to any directory in the Archive app, in addition to the default behavior of opening a file in it's default app.

![ScreenRecording2024-12-25at6 19 41AM-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/60cb2849-a654-453e-9ea6-37461c0c210e)

-----
# For the reviewer

- [x] I've uploaded the firmware with this patch to a device and verified its functionality
- [x] I've confirmed the bug to be fixed / feature to be stable

956MB committed:
feat(Desktop): Directories support for keybinds

- Adds *RIGHT* button select in the file browser dialogs and changing the `Open File` action to `Open File/Directory` in `Settings > Desktop > Keybinds Setup`. This adds the ability to open to any directory in the Archive app, in addition to the default behavior of opening a file in it's default app.

956MB committed:
line order mixup

956MB committed:
Main Menu: Allow adding JS files (or any file)

- Normal files and directories are now able to be added to then main menu and are run in their appropriate apps.
- e.g. .txt files shown in text viewer, .js files are run in the JS Runner app, and folders are navigated to by the Archive app. All similar to the desktop keybinds functionality.
- Icons are also assigned appropriately based on the extensions, though more could probably be added to the `loader_menu_get_ext_icon` function.
- Also replaced some of the long arduous is_dir checks and just used the `storage_dir_exists` function since its already there and does the same.

956MB commented:
The adding of JS files and other stuff to the main menu seemed connected to the initial PR with the `select_right` thing for the keybinds dialog, so I just added on to this PR instead of creating a new one for this feature.

### Screenshot of the files/folder with icons in the Main Menu

![Screenshot-20241227-205546](https://github.com/user-attachments/assets/e555a36c-0f42-436d-a148-02ea8f7cdd93)

Another thing I was thinking about but haven't messed with is whether the word `App` ought to still be used in the menu, as it can basically have any file type with this change, not just apps. For simplicity sake it'd probably be best just to keep it, though I think `File` can work well if you think of everything on the device as 'a file', including apps.

- "App 1/12" -> "File 1/12"
- "Add App" -> "Add File"
- "Move App" -> "Move FIle"
- "Remove App" -> "Remove File"

Just a thought.

WillyJL commented:
> The adding of JS files and other stuff to the main menu seemed connected to the initial PR with the `select_right` thing for the keybinds dialog, so I just added on to this PR instead of creating a new one for this feature.

very nice!



> Another thing I was thinking about but haven't messed with is whether the word `App` ought to still be used in the menu, as it can basically have any file type with this change, not just apps. For simplicity sake it'd probably be best just to keep it, though I think `File` can work well if you think of everything on the device as 'a file', including apps.
> 
> * "App 1/12" -> "File 1/12"
> * "Add App" -> "Add File"
> * "Move App" -> "Move FIle"
> * "Remove App" -> "Remove File"
> 
> Just a thought.

i think "Item" would work better as a single name for both apps and files. or alternatively, could use both app and file, changing the text of the buttons based on what is currently selected, but that is a bit over the top i think

956MB commented:
> i think "Item" would work better as a single name for both apps and files.

Oh yep, you're right. "Item" is generic enough for all of it 👍

956MB committed:
should be checking `ext` for NULL

WillyJL commented on applications/services/loader/loader_menu.c:
```diff
@@ -153,7 +155,28 @@ static void loader_menu_apps_callback(void* context, uint32_t index) {
+        furi_record_close(RECORD_STORAGE);
+
+        if(is_dir) {
+            Desktop* desktop = furi_record_open(RECORD_DESKTOP);
+            desktop->archive_dir = furi_string_alloc_set(menu_app->path);
+            furi_record_close(RECORD_DESKTOP);
+
+            furi_thread_flags_set(furi_thread_get_id(app->loader_menu->thread), 0);
+            if(app->loader_menu->closed_cb) app->loader_menu->closed_cb(app->loader_menu->context);
+
+            view_dispatcher_send_custom_event(
+                desktop->view_dispatcher, DesktopMainEventOpenArchive);
```
i get why its like this, but i hate it xD

ill try to come up with a way to launch archive without passing through desktop

WillyJL replied:
reworked it into desktop with less FuriStrings being passed around, and also handled folders in run_with_default_app() for less duplicated code

WillyJL committed:
Merge remote-tracking branch 'mntm/dev' into mntm-pr-331

WillyJL committed:
Move select_right at end of structs for binary compatibility

apps may blindly reach into these structs so need to keep the basics in same structure

for DialogsFileBrowserOptions this is even in public api and after compilation this would be incompatible with other firmwares even without reaching into private structs

WillyJL committed:
Select menu item / folder for directories too

WillyJL committed:
Move api below too

WillyJL committed:
Keep ofw order here too

WillyJL committed:
Refactor starting archive into desktop, less FuriString passing around

WillyJL committed:
Dont leave main menu when launching archive

WillyJL committed:
Simplify/fix a few things

WillyJL committed:
Handle folders in run_with_default_app()

WillyJL committed:
Update App -> Item naming in MNTM settings

WillyJL committed:
Fix build

WillyJL reviewed (approved):
*one eternity later*

WillyJL committed:
Explain pressing right

WillyJL committed:
Update changelog

WillyJL committed:
Merge remote-tracking branch 'mntm/dev' into mntm-pr-331

956MB commented:
Merged :) #369
